### ISSUE TITLE
Unsafe close of outbound channel

### ISSUE THREAD

**Issue by yh742** [mochi-co/mqtt#202]
Hi there,

Found a possible issue while running test (related to #193). It seems like closing the outbound channel can cause race condition since other publisher could be publishing to it at the same time.

```
WARNING: DATA RACE
Write at 0x00c00048c070 by goroutine 21:
  runtime.closechan()
      /usr/local/go/src/runtime/chan.go:357 +0x0
  github.com/mochi-co/mqtt/v2.(*Client).Stop.func1()
      /home/sean/go/pkg/mod/github.com/mochi-co/mqtt/v2@v2.2.7/clients.go:368 +0x10f
  sync.(*Once).doSlow()
      /usr/local/go/src/sync/once.go:74 +0x101
  sync.(*Once).Do()
      /usr/local/go/src/sync/once.go:65 +0x46
  github.com/mochi-co/mqtt/v2.(*Client).Stop()
      /home/sean/go/pkg/mod/github.com/mochi-co/mqtt/v2@v2.2.7/clients.go:358 +0x74
  github.com/mochi-co/mqtt/v2.(*Server).processDisconnect()
      /home/sean/go/pkg/mod/github.com/mochi-co/mqtt/v2@v2.2.7/server.go:1150 +0x1bb
  github.com/mochi-co/mqtt/v2.(*Server).processPacket()
      /home/sean/go/pkg/mod/github.com/mochi-co/mqtt/v2@v2.2.7/server.go:561 +0x787
  github.com/mochi-co/mqtt/v2.(*Server).receivePacket()
      /home/sean/go/pkg/mod/github.com/mochi-co/mqtt/v2@v2.2.7/server.go:411 +0x87
  github.com/mochi-co/mqtt/v2.(*Server).receivePacket-fm()
      <autogenerated>:1 +0x89
  github.com/mochi-co/mqtt/v2.(*Client).Read()
      /home/sean/go/pkg/mod/github.com/mochi-co/mqtt/v2@v2.2.7/clients.go:349 +0x19c
  github.com/mochi-co/mqtt/v2.(*Server).attachClient()
      /home/sean/go/pkg/mod/github.com/mochi-co/mqtt/v2@v2.2.7/server.go:366 +0x9c4
  github.com/mochi-co/mqtt/v2.(*Server).EstablishConnection()
      /home/sean/go/pkg/mod/github.com/mochi-co/mqtt/v2@v2.2.7/server.go:312 +0x88
  github.com/mochi-co/mqtt/v2.(*Server).EstablishConnection-fm()
      <autogenerated>:1 +0x6d
  github.com/mochi-co/mqtt/v2/listeners.(*TCP).Serve.func1()
      /home/sean/go/pkg/mod/github.com/mochi-co/mqtt/v2@v2.2.7/listeners/tcp.go:84 +0x8c

Previous read at 0x00c00048c070 by goroutine 12:
  runtime.chansend()
      /usr/local/go/src/runtime/chan.go:160 +0x0
  github.com/mochi-co/mqtt/v2.(*Server).publishToClient()
      /home/sean/go/pkg/mod/github.com/mochi-co/mqtt/v2@v2.2.7/server.go:854 +0xa94
  github.com/mochi-co/mqtt/v2.(*Server).publishToSubscribers()
      /home/sean/go/pkg/mod/github.com/mochi-co/mqtt/v2@v2.2.7/server.go:785 +0x604
  github.com/mochi-co/mqtt/v2.(*Server).processPublish()
      /home/sean/go/pkg/mod/github.com/mochi-co/mqtt/v2@v2.2.7/server.go:748 +0xba4
  github.com/mochi-co/mqtt/v2.(*Server).processPacket()
      /home/sean/go/pkg/mod/github.com/mochi-co/mqtt/v2@v2.2.7/server.go:569 +0x2a7
  github.com/mochi-co/mqtt/v2.(*Server).receivePacket()
      /home/sean/go/pkg/mod/github.com/mochi-co/mqtt/v2@v2.2.7/server.go:411 +0x87
  github.com/mochi-co/mqtt/v2.(*Server).receivePacket-fm()
      <autogenerated>:1 +0x89
  github.com/mochi-co/mqtt/v2.(*Client).Read()
      /home/sean/go/pkg/mod/github.com/mochi-co/mqtt/v2@v2.2.7/clients.go:349 +0x19c
  github.com/mochi-co/mqtt/v2.(*Server).attachClient()
      /home/sean/go/pkg/mod/github.com/mochi-co/mqtt/v2@v2.2.7/server.go:366 +0x9c4
  github.com/mochi-co/mqtt/v2.(*Server).EstablishConnection()
      /home/sean/go/pkg/mod/github.com/mochi-co/mqtt/v2@v2.2.7/server.go:312 +0x88
  github.com/mochi-co/mqtt/v2.(*Server).EstablishConnection-fm()
      <autogenerated>:1 +0x6d
  github.com/mochi-co/mqtt/v2/listeners.(*TCP).Serve.func1()
      /home/sean/go/pkg/mod/github.com/mochi-co/mqtt/v2@v2.2.7/listeners/tcp.go:84 +0x8c

```

**Comment by thedevop** [comment]
@mochi-co ,

The only reason outbound channel needs to be closed is the WriteLoop is using range, and that Go routine needs to end when the client is stopped.

If we change ClientState.done to a context. That way, WriteLoop can use something like below, :
```go
	for {
		select {
		case pk := <-cl.State.outbound:
			if err := cl.WritePacket(*pk); err != nil {
				cl.ops.log.Debug().Err(err).Str("client", cl.ID).Interface("packet", pk).Msg("failed publishing packet")
			}
			atomic.AddInt32(&cl.State.outboundQty, -1)
		case <-cl.State.done.Done():
			return
		}
	}
```

For places where we check the state, it can be changed to cl.State.done.Err().

Alternative, if we introduce lock for #196, we may use the same mechanism.

**Comment by mochi-co** [comment]
I added a lock around this #213 in conjunction with #212 which should theoretically fix it. I've merged these changes into v2.2.8, but please reopen if it's not fixed.

**Comment by thedevop** [comment]
@mochi-co , this problem should still exist, as the producer into the channel doesn't obtain the lock and check state before pushing into the channel.
https://github.com/mochi-co/mqtt/blob/c274d5fd08456605ea197a92778a55bb57db16cb/server.go#L853-L856

The more I think about it, I think it's better to change the State.done to a context.

**Comment by mochi-co** [comment]
@thedevop I've replaced cl.State.done with a cl.State.open context, as you suggested 👍🏻 
@yh742 This has been implemented in v2.2.9, let me know if it resolves the issue, otherwise reopen the issue :)
